src/metaopt/io/database/__init__.py

```diff
@@ -0,0 +1,121 @@
+# -*- coding: utf-8 -*-
+"""
+:mod:`metaopt.io.database` -- Wrappers for database frameworks
+==============================================================
+
+.. module:: database
+   :platform: Unix
+   :synopsis: Import name for wrappers of db frameworks, like MongoDB
+
+Contains :class:`AbstractDB`, an interface for databases.
+Currently, implemented wrappers:
+
+   - :class:`metaopt.io.database.mongodb.MongoDB`
+
+"""
+from __future__ import absolute_import
+
+from abc import abstractmethod, abstractproperty
+import logging
+
+import six
+
+from metaopt.utils import AbstractSingletonType
+
+
+@six.add_metaclass(AbstractSingletonType)
+class AbstractDB(object):
+    """Base class for database framework wrappers.
+
+    Attributes
+    ----------
+    uri : str
+       Database framework specific connection string.
+    dbname : str
+       Name of database containing experiments.
+    host : str
+       Known hostname or IP address in which database server resides.
+    port : int
+       Port that database server listens to for requests.
+    username : str
+        Name of user with write/read permissions to database with name `dbname`.
+    password : str
+        Secret phrase of user, `username`.
+
+    """
+
+    def __init__(self, uri=None, dbname='metaopt', host='localhost',
```

Comment:
I love it. Defining the attributes separately in the configuration file is more natural than the URI. The URI is more convenient for passing as an argument in command line.

So as you propose I would make the `host`/`uri` the first and only positional argument and all the arguments would be passed to MongoClient rather than the URI. That would also address Pascal's concern about the username/password in the URI. In that case, I would drop everything related to URI; the property and self._uri. Until we add support for URI, MongoDB("some_uri") would fail.